sdk/attestation/Azure.Security.Attestation/src/AttestationToken.cs

```diff
@@ -143,43 +188,309 @@ protected AttestationToken()
+                        HashAlgorithmName.SHA256);
+                    if (signatureValidated)
+                    {
+                        actualSigner = signer;
+                    }
+                    break;
+                }
+            }
+
+            if (actualSigner != null)
+            {
+                this.CertificateThumbprint = actualSigner.SigningCertificates[0].Thumbprint;
+                this.SigningCertificate = actualSigner;
+            }
+            return actualSigner != null;
+        }
+
+        /// <summary>
+        /// Validate the common properties of the attestation token - ensure that the expiration time for the token is "reasonable", if present.
+        /// </summary>
         /// <returns></returns>
-        public virtual bool ValidateToken(IReadOnlyList<AttestationSigner> attestationSigningCertificates, Func<AttestationToken, AttestationSigner, bool> validationCallback = default)
+        private bool ValidateCommonProperties(TokenValidationOptions options)
         {
-            Argument.AssertNotNull(attestationSigningCertificates, nameof(attestationSigningCertificates));
-            Argument.AssertNotNull(attestationSigningCertificates[0], nameof(attestationSigningCertificates));
-            Argument.AssertNotNullOrEmpty(attestationSigningCertificates, nameof(attestationSigningCertificates));
+            DateTimeOffset timeNow = DateTimeOffset.Now;
+            if (Payload.ExpirationTime.HasValue && (options?.ValidateExpirationTime ?? true))
+            {
+                if (timeNow.CompareTo(ExpirationTime.Value) > 0)
+                {
+                    if (options?.TimeValidationSlack != 0)
+                    {
+                        var delta = timeNow.Subtract(ExpirationTime.Value);
+                        if (delta.Seconds > options?.TimeValidationSlack)
+                        {
+                            throw new Exception("Attestation token expired before the slack time provided.");
+                        }
+                    }
+                    else
+                    {
+                        throw new Exception("Attestation token expired.");
+                    }
+                }
+            }
 
-            if (validationCallback != null)
+            if (Payload.NotBeforeTime.HasValue && (options?.ValidateNotBeforeTime?? true))
             {
-                return validationCallback(this, attestationSigningCertificates[0]);
+                if (DateTimeOffset.Now.CompareTo(NotBeforeTime.Value) < 0)
+                {
+                    if (options?.TimeValidationSlack != 0)
+                    {
+                        var delta = timeNow.Subtract(NotBeforeTime.Value);
+                        if (Math.Abs(delta.Seconds) > options?.TimeValidationSlack)
+                        {
+                            throw new Exception("Attestation token will not become valid before the slack time provided.");
```

Comment:
I asked the teams channel about a better exception but had no responses. I *hate* using Exception here, but I don't know what a better exception is. And there is strong guidance about creating your own exception class.

Reply:
That's entirely fair.  I don't have a really great suggestion either.   I'd say `InvalidOperationException` to denote that the state of the token makes it impossible to perform the requested operation calling for validation, but since this is a public type, that kind of breaks down.

Would customers want to detect this scenario and potentially take corrective action?   That would be justification to discuss a custom exception with the architects.   If not, I.... don't really have anything better than `Exception` to offer.

Reply:
There's really no action that the customer can take - it's an indication that the token was invalid (similar to an AAD bearer token with a signature failure). This is pretty catastrophic from a customer standpoint - they got something from the server which wasn't valid. But ultimately there's no action they can take - it's just horrible.